certbot-apache/certbot_apache/_internal/configurator.py

```diff
@@ -225,6 +237,42 @@ def updated_mod_ssl_conf_digest(self):
         """Full absolute path to digest of updated SSL configuration file."""
         return os.path.join(self.config.config_dir, constants.UPDATED_MOD_SSL_CONF_DIGEST)
 
+    def _open_module_file(self, ssl_module_location):
+        """Extract the open lines of openssl_version for testing purposes"""
+        try:
+            with open(ssl_module_location, mode="rb") as f:
+                contents = f.read()
+        except IOError:
+            return None
+        return contents
+
+    def openssl_version(self):
+        """Lazily retrieve openssl version"""
+        if self._openssl_version:
+            return self._openssl_version
+        # Step 1. Check for LoadModule directive
+        try:
+            ssl_module_location = self.parser.modules['ssl_module']
+        except KeyError:
+            logger.warning("Could not find ssl_module; not disabling session tickets.")
```

Comment:
I think you're probably right that we want to warn in this case too if we're going to error if `self.parser.modules['ssl_module']` is `None`, however, this has the effect of printing this warning on fresh Debian/Ubuntu systems when we haven't enabled the SSL module yet. I don't think we should do this.

To avoid this, I think we essentially need a way to check if we should assume of mod SSL is available yet. A couple ideas here are  just checking if `self.option("handle_modules")` is true in this except block or adding another argument to `install_ssl_options_conf` that's passed all the way here. What do you think?

Reply:
Ugh, I honestly hate adding contextual checks to utility methods. So annoying to trace what we expect to run when reading it later. Here's an alternate proposal: we don't warn on this. Because `enable_mod` [raises an error](https://github.com/certbot/certbot/blob/4abd81e2186eddc67551d61a8260440bd177d18d/certbot-apache/certbot_apache/_internal/configurator.py#L2190) if `ssl_module` isn't found. So, when he hit this line, either a) we expected it or b) something is [terribly wrong](https://www.sbnation.com/a/17776-football/). So what if we just return here and don't print anything? Basically this check will be a switch to know which of the two states we're in. I'm happy to add an explanation to this effect.

Reply:
Good idea! I also like that idea better.

We can also be sure it's in the list of modules after the override for Debian too since we manually add it at https://github.com/certbot/certbot/blob/9e5bca4bbf3c4fd198382dd9d6e05b54014aca30/certbot-apache/certbot_apache/_internal/override_debian.py#L121

I think adding a comment so we remember this stuff in the future is a great idea too.

In looking into this, I think I found another problem though. When the Debian override adds the SSL module to the modules list, it doesn't add the path to the shared object file (see the above link) and as a result the second time `openssl_version` is called we hit the case where `self.parser.modules['ssl_module']` is `None` and as a result still print a warning and do not disable session tickets on a fresh Debian/Ubuntu install. I think we should probably fix this too.

Reply:
This comment is a little more hastly written than normal, but I wanted to flag this now in case you kept working on this tonight.

I think I may have thought of a problem with your plan to handle Debian and `enable_mod`. We only call `enable_mod("ssl")` on systems where `self.option("handle_modules")` is true (which I think only Debian off the top of my head). See https://github.com/certbot/certbot/blob/3bfaf41d3df0417a2d58d088af1a893d320aeb55/certbot-apache/certbot_apache/_internal/configurator.py#L1238-L1243

If that's right, then I think there's a realistic option (c) to hitting this case which is you're on a non-Debian system and we're failing to parse your modules for whatever reason, but it's not an error because we essentially assume that your modules are set up correctly.

Reply:
I knew that it was setting it to `None` and thought I accounted for that but can't remember how.

Reply:
Oh, I think it was by making sure that `parse_modules` is called again, after the `add_mod` call, but maybe that's not actually the case. Because I had an earlier implementation where I tried to set it to something other than `None`, but then decided that calling `parse_modules` again is easier.

Reply:
This addition here: https://github.com/certbot/certbot/pull/7771/files#diff-2caba5786389c0d032404485e9936206R1293

Reply:
> If that's right, then I think there's a realistic option (c) to hitting this case which is you're on a non-Debian system and we're failing to parse your modules for whatever reason, but it's not an error because we essentially assume that your modules are set up correctly.

This is a good point. But I also think that we're fine not logging anything here; if `ssl_module` isn't installed, there's no way certbot works, so we'll error somewhere else to fix it, no? Like, printing that we're not disabling session tickets is nice, but probably unnecessary relative to the other issues that are going to need to be fixed in that case.

Reply:
Oh, I see what you mean about `self.option("handle_modules")` now; the github interface was cutting that off. Looking again.